### PR TITLE
Dev ngc-learn merged with ngc-learn

### PR DESCRIPTION
**Pull request by ago109**
This is a merger of the dev branch of ngc-learn and ngc-learn -- the heavily revised ngc-learn now supports a cleaner interface for crafting models/systems of components (with the updated nodes-and-cables system); the simulation backend is also significantly faster, facilitating global jit-i-fication and function purification:
* the ngc-sim-lib backend now features an in-built compiler for reading in model components and wiring patterns; this ensures everything is compliant with how jax's just-in-time compiler operates; loop scanning compilation is now supported (e.g., globally jit-i-fying spiking net processing loops over a time window of length T)
* the adaptive exponential integrator unit has been written and integrated
* all components (neurons, other, input-encoders, synapses) rewritten in terms of new ngc-sim-lib/nodes-and-cables compilation backend
* new tutorials written and old tutorials re-written to fit with new format
* model museum models re-written to comply with new format

### CODE REVIEW

File: ngclearn/components/neurons/graded/laplacianErrorCell.py [hunk 1]
```diff
@@ -60,126 +65,73 @@ class LaplacianErrorCell(Component): ## Rate-coded/real-valued error unit/cell
...
-        self.mean = jnp.zeros((self.batch_size, self.n_units)) #None
-        self.modulator = None
-
-    def save(self, **kwargs):
-        pass
+
+        ## Compartment setup
+        restVals = jnp.zeros((self.batch_size, self.n_units))
+        self.j = Compartment(None) # ## electrical current/ input compartment/to be wired/set. # NOTE: VN: This is never used
+        self.L = Compartment(None) # loss compartment
```
Comment (willgebhardt) [inline ngclearn/components/neurons/graded/laplacianErrorCell.py:81]: This needs to be the default value to work with scanner

File: ngclearn/components/input_encoders/bernoulliCell.py [hunk 2]
```diff
@@ -48,76 +53,70 @@ class BernoulliCell(Component):
...
-
-    def reset(self, **kwargs):
-        self.inputCompartment = None
-        self.outputCompartment = jnp.zeros((self.batch_size, self.n_units)) #None
-        self.timeOfLastSpike = jnp.zeros((self.batch_size, self.n_units))
 
-    def save(self, **kwargs):
-        pass
+        # Compartments (state of the cell, parameters, will be updated through stateless calls)
+        self.inputs = Compartment(None) # input compartment
```
Comment (willgebhardt) [inline ngclearn/components/input_encoders/bernoulliCell.py:67]: Needs a constant shape to work with scanners. (Either always none or a shaped array)

File: ngclearn/components/input_encoders/latencyCell.py [hunk 3]
```diff
@@ -203,47 +163,112 @@ def __init__(self, name, n_units, tau=1., threshold=0.01, first_spike_time=0.,
...
-        self.inputCompartment = None
-        self.outputCompartment = jnp.zeros((self.batch_size, self.n_units)) #None
-        self.timeOfLastSpike = jnp.zeros((self.batch_size, self.n_units))
-        self._mask = jnp.zeros((self.batch_size, self.n_units))
-
-    def save(self, **kwargs):
-        pass
+
+        ## Compartment setup
+        self.inputs = Compartment(None) # input compartment
```
Comment (willgebhardt) [inline ngclearn/components/input_encoders/latencyCell.py:171]: Needs a constant shape to work with scanners

File: ngclearn/components/input_encoders/latencyCell.py [hunk 4]
```diff
@@ -203,47 +163,112 @@ def __init__(self, name, n_units, tau=1., threshold=0.01, first_spike_time=0.,
...
+            targ_sp_times = stimes #* calcEvent + targ_sp_times * (1. - calcEvent)
+        else: ## standard nonlinear spike time calculation
+            stimes = calc_spike_times_nonlinear(data, tau, threshold,
+                                                first_spike_time,
+                                                num_steps=num_steps,
+                                                normalize=normalize)
+            targ_sp_times = stimes #* calcEvent + targ_sp_times * (1. - calcEvent)
+        return targ_sp_times
+
+    @resolver(_calc_spike_times, output_compartments=['targ_sp_times'])
```
Comment (willgebhardt) [inline ngclearn/components/input_encoders/latencyCell.py:196]: Unroll the output compartments

File: ngclearn/components/input_encoders/poissonCell.py [hunk 5]
```diff
@@ -56,79 +60,73 @@ class PoissonCell(Component):
...
-
-    def reset(self, **kwargs):
-        self.inputCompartment = None
-        self.outputCompartment = jnp.zeros((self.batch_size, self.n_units)) #None
-        self.timeOfLastSpike = jnp.zeros((self.batch_size, self.n_units))
 
-    def save(self, **kwargs):
-        pass
+        ## Compartment setup
+        self.inputs = Compartment(None) # input compartment
```
Comment (willgebhardt) [inline ngclearn/components/input_encoders/poissonCell.py:77]: Needs a constant shape to work with scanners.

File: ngclearn/components/neurons/graded/gaussianErrorCell.py [hunk 6]
```diff
@@ -60,126 +65,73 @@ class GaussianErrorCell(Component): ## Rate-coded/real-valued error unit/cell
...
-        self.reset()
-
-    def verify_connections(self):
-        self.metadata.check_incoming_connections(self.meanName(), min_connections=1)
-        self.metadata.check_incoming_connections(self.targetName(), min_connections=1)
 
-    def advance_state(self, t, dt, **kwargs):
+        ##Random Number Set up
+        restVals = jnp.zeros((self.batch_size, self.n_units))
+        self.j = Compartment(None) # ## electrical current/ input compartment/to be wired/set. # NOTE: VN: This is never used
```
Comment (willgebhardt) [inline ngclearn/components/neurons/graded/gaussianErrorCell.py:78]: remove unused compartment

File: ngclearn/components/neurons/graded/gaussianErrorCell.py [hunk 7]
```diff
@@ -60,126 +65,73 @@ class GaussianErrorCell(Component): ## Rate-coded/real-valued error unit/cell
...
+    @staticmethod
+    def _reset(batch_size, n_units):
+        dmu = jnp.zeros((batch_size, n_units))
+        dtarget = jnp.zeros((batch_size, n_units))
+        target = jnp.zeros((batch_size, n_units)) #None
+        mu = jnp.zeros((batch_size, n_units)) #None
+        return dmu, dtarget, target, mu
+
+    @resolver(_reset)
+    def reset(self, dmu, dtarget, target, mu):
```
Comment (willgebhardt) [inline ngclearn/components/neurons/graded/gaussianErrorCell.py:111]: This reset call will not be mirror inside of a compiled function due to their being runtime logic outside of the resolved static method

File: ngclearn/components/neurons/graded/laplacianErrorCell.py [hunk 8]
```diff
@@ -60,126 +65,73 @@ class LaplacianErrorCell(Component): ## Rate-coded/real-valued error unit/cell
...
-        self.target = jnp.zeros((self.batch_size, self.n_units)) #None
-        self.mean = jnp.zeros((self.batch_size, self.n_units)) #None
-        self.modulator = None
-
-    def save(self, **kwargs):
-        pass
+
+        ## Compartment setup
+        restVals = jnp.zeros((self.batch_size, self.n_units))
+        self.j = Compartment(None) # ## electrical current/ input compartment/to be wired/set. # NOTE: VN: This is never used
```
Comment (willgebhardt) [inline ngclearn/components/neurons/graded/laplacianErrorCell.py:80]: remove unused compartments

File: ngclearn/components/neurons/graded/laplacianErrorCell.py [hunk 9]
```diff
@@ -60,126 +65,73 @@ class LaplacianErrorCell(Component): ## Rate-coded/real-valued error unit/cell
...
+    @staticmethod
+    def _reset(batch_size, n_units):
+        dmu = jnp.zeros((batch_size, n_units))
+        dtarget = jnp.zeros((batch_size, n_units))
+        target = jnp.zeros((batch_size, n_units)) #None
+        mu = jnp.zeros((batch_size, n_units)) #None
+        return dmu, dtarget, target, mu
+
+    @resolver(_reset)
+    def reset(self, dmu, dtarget, target, mu):
```
Comment (willgebhardt) [inline ngclearn/components/neurons/graded/laplacianErrorCell.py:112]: This reset function will not mirror behavior inside of compiled functions due to runtime logic existing outside the resolved static method

File: ngclearn/components/neurons/spiking/izhikevichCell.py [hunk 10]
```diff
@@ -278,39 +217,72 @@ def __init__(self, name, n_units, tau_m=1., R_m=1., v_thr=30., v_reset=-65.,
...
+    @staticmethod
+    def _advance_state(t, dt, tau_m, tau_w, v_thr, coupling, v_reset, w_reset, R_m,
+                     intgFlag, j, v, w, s, tols):
+        v, w, s = run_cell(dt, j, v, s, w, v_thr=v_thr, tau_m=tau_m, tau_w=tau_w,
+                           b=coupling, c=v_reset, d=w_reset, R_m=R_m,
+                           integType=intgFlag)
+        tols = update_times(t, s, tols)
+        return j, v, w, s, tols
+
+    @resolver(_advance_state, output_compartments=['j', 'v', 'w', 's', 'tols'])
```
Comment (willgebhardt) [inline ngclearn/components/neurons/spiking/izhikevichCell.py:242]: unroll output compartments inside of the function arguments

File: ngclearn/components/other/expKernel.py [hunk 11]
```diff
@@ -1,123 +1,110 @@
...
+        self.tau_w = tau_w ## kernel window time constant
+        self.nu = nu
+        self.win_len = int(nu/dt) + 1 ## window length
+        #tf ## window of spike times
+
+        ## Layer Size Setup
+        self.batch_size = 1
+        self.n_units = n_units
+
+        self.inputs = Compartment(None) # input comp
```
Comment (willgebhardt) [inline ngclearn/components/other/expKernel.py:61]: Needs a constant shape to work with scanners.

File: ngclearn/components/other/varTrace.py [hunk 12]
```diff
@@ -56,89 +61,76 @@ class VarTrace(Component): ## low-pass filter
...
-
-    def reset(self, **kwargs):
-        self.trace = jnp.zeros((1, self.n_units))
-        self.outputCompartment = self.trace
-        self.inputCompartment = None
-
-    def save(self, **kwargs):
-        pass
+
+        self.inputs = Compartment(None) # input compartment
```
Comment (willgebhardt) [inline ngclearn/components/other/varTrace.py:83]: Needs a constant shape to work with scanners.

File: ngclearn/components/synapses/__init__.py [hunk 13]
```diff
@@ -1,3 +1,4 @@
 from .hebbian.hebbianSynapse import HebbianSynapse
 from .hebbian.traceSTDPSynapse import TraceSTDPSynapse
 from .hebbian.expSTDPSynapse import ExpSTDPSynapse
+#from .hebbian.eventSTDPSynapse import EventSTDPSynapse
```
Comment (willgebhardt) [inline ngclearn/components/synapses/__init__.py:4]: remove

File: ngclearn/components/synapses/hebbian/__init__.py [hunk 14]
```diff
@@ -1,3 +1,4 @@
 from .hebbianSynapse import HebbianSynapse
 from .traceSTDPSynapse import TraceSTDPSynapse
 from .expSTDPSynapse import ExpSTDPSynapse
+#from .eventSTDPSynapse import EventSTDPSynapse
```
Comment (willgebhardt) [inline ngclearn/components/synapses/hebbian/__init__.py:4]: remove

File: ngclearn/utils/__init__.py [hunk 15]
```diff
@@ -0,0 +1,29 @@
+import numpy as np
+import jax
+import jax.numpy as jnp
+
+def tensorstats(tensor):
```
Comment (willgebhardt) [inline ngclearn/utils/__init__.py:5]: why is this not in a file and lives in an init?

File: ngclearn/utils/optim/__init__.py [hunk 16]
```diff
@@ -1,2 +1,20 @@
-from .sgd import SGD
-from .adam import Adam
+import functools
+
+from .sgd import sgd_step, sgd_init
+from .adam import adam_step, adam_init
+
+def get_opt_init_fn(opt='adam'):
```
Comment (willgebhardt) [inline ngclearn/utils/optim/__init__.py:6]: why are there functions defined in an init and not in a regular file?

**Review by willgebhardt — CHANGES_REQUESTED** [review]
Please address the comments provided to proceed with the merge

File: ngclearn/utils/optim/__init__.py (hunk 16 above)
Comment (ago109) [reply ngclearn/utils/optim/__init__.py:6]: Viet - frowny face

**Review by willgebhardt — APPROVED** [review]
Looks good

File: ngclearn/utils/__init__.py (hunk 15 above)
Comment (rxng8) [reply ngclearn/utils/__init__.py]: I just don't know where to put it.

Comment (rxng8) [reply ngclearn/utils/__init__.py]: @willgebhardt which file do you recommend me putting in, I will make necessary refactoring changes

File: ngclearn/utils/optim/__init__.py (hunk 16 above)
Comment (rxng8) [reply ngclearn/utils/optim/__init__.py]: Well, same thought on the last comment. For me, it's totally fine to do so. But yes, please suggest a file name and I'll make necessary refactoring changes

**Pull request merged**